File: Week 27/cpp/sweep_line.cpp
/*
 * WEEK 27 - C++ ADVANCED TOPICS
 * Topic: Sweep Line - Skyline & Max Overlap
 * File: sweep_line.cpp
 *
 * CONCEPT:
 *   Sweep an imaginary vertical line across the plane, processing events in
 *   x order while maintaining an "active set". Two showcase problems:
 *     - Skyline: max-heap of active building heights gives the silhouette.
 *     - Max overlap: counting +1/-1 events finds the maximum simultaneous
 *       active intervals.
 *
 * KEY POINTS:
 *   - Sort events by primary x then a sensible tie-breaker.
 *   - For skyline: use a max-heap (priority_queue<pair<int,int>>) of
 *     (height, end_x); pop expired tops lazily.
 *
 * ALGORITHM / APPROACH:
 *   SKYLINE: events for each building -> (l, h, r) start, (r, 0, 0) end.
 *           sort; max-heap of (h, r); on each event update heap, lazily
 *           drop expired, emit (x, h_max) when h_max changes.
 *   OVERLAP: events (start,+1), (end,-1); sort; running counter -> max.
 *
 * C++-SPECIFIC NOTES:
 *   - priority_queue<pair<int,int>> is a max-heap by default.
 *   - Use INT_MAX as a sentinel "never expires" entry to avoid empty-heap
 *     edge cases.
 *
 * DRY RUN / EXAMPLE:
 *   buildings = {{2,9,10},{3,7,15},{5,12,12},{15,20,10},{19,24,8}}
 *   skyline   = {(2,10),(3,15),(7,12),(12,0),(15,10),(20,8),(24,0)}
 *
 * COMPLEXITY:
 *   Time O(n log n), Space O(n).
 */

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>

using namespace std;

vector<pair<int,int>> skyline(vector<tuple<int,int,int>> buildings) {
    // events: (x, height) — height>0 = start, ==0 = end
    vector<tuple<int,int,int>> events; // (x, h, r)
    for (auto& [l, r, h] : buildings) {
        events.emplace_back(l, h, r);     // start
        events.emplace_back(r, 0, 0);     // end
    }
    sort(events.begin(), events.end(), [](auto& a, auto& b) {
        if (get<0>(a) != get<0>(b)) return get<0>(a) < get<0>(b);
        return get<1>(a) > get<1>(b); // taller starts first
    });
    priority_queue<pair<int,int>> heap; // (height, end_x)
    heap.emplace(0, INT_MAX);
    vector<pair<int,int>> result;
    for (auto& [x, h, r] : events) {
        if (h) heap.emplace(h, r);
        while (heap.top().second <= x) heap.pop();
        int curH = heap.top().first;
        if (result.empty() || result.back().second != curH)
            result.emplace_back(x, curH);
    }
    return result;
}

int maxOverlap(vector<pair<int,int>> intervals) {
    vector<pair<int,int>> events;
    for (auto& [s, e] : intervals) {
        events.emplace_back(s, +1);
        events.emplace_back(e, -1);
    }
    sort(events.begin(), events.end(), [](auto& a, auto& b) {
        if (a.first != b.first) return a.first < b.first;
        return a.second > b.second; // opens before closes at same x
    });
    int cur = 0, best = 0;
    for (auto& [x, d] : events) {
        cur += d;
        best = max(best, cur);
    }
    return best;
}

int main() {
    auto sk = skyline({{2,9,10},{3,7,15},{5,12,12},{15,20,10},{19,24,8}});
    cout << "Skyline:";
    for (auto& [x, h] : sk) cout << " (" << x << "," << h << ")";
    cout << "\n";
    cout << "Max overlap: " << maxOverlap({{1,5},{2,6},{3,4},{7,8}}) << "\n";
    return 0;
}

/*
 * NOTES
 * -----
 * Differences from Java:
 *   - Java's geometry.java does not include sweep-line algorithms; we add
 *     the skyline and overlap classics for completeness.
 *   - std::priority_queue + lazy deletion (sentinel INT_MAX) avoids the
 *     overhead of a full balanced BST.
 */

File: Week 27/rust/sweep_line.rs
// WEEK 27 - RUST ADVANCED TOPICS
// Topic: Sweep Line - Skyline & Max Overlap
// File: sweep_line.rs
//
// CONCEPT:
//   Sweep an imaginary vertical line across the plane, processing events in
//   x order while maintaining an active set. Two classics:
//     - Skyline: max-heap of building heights gives the silhouette.
//     - Max overlap: counting +1/-1 events finds the simultaneous maximum.
//
// KEY POINTS:
//   - Sort events lexicographically (x then tie-breaker).
//   - For skyline: BinaryHeap of (height, end_x); lazy pop expired tops.
//
// ALGORITHM / APPROACH:
//   skyline: events for (l, r, h) -> (l, h, r) start, (r, 0, 0) end.
//            sort; max-heap on h; emit (x, h_max) whenever h_max changes.
//   overlap: +1 on start, -1 on end; sort with +1 before -1 at same x.
//
// RUST-SPECIFIC NOTES:
//   - std::collections::BinaryHeap is a max-heap by default.
//   - Use std::cmp::Reverse for min-heap behaviour if needed.
//   - Use i32 throughout; convert at boundaries.
//
// DRY RUN / EXAMPLE:
//   buildings = [(2,9,10),(3,7,15),(5,12,12),(15,20,10),(19,24,8)]
//   skyline   = [(2,10),(3,15),(7,12),(12,0),(15,10),(20,8),(24,0)]
//
// COMPLEXITY:
//   Time O(n log n)   Space O(n)

use std::collections::BinaryHeap;

pub fn skyline(buildings: &[(i32, i32, i32)]) -> Vec<(i32, i32)> {
    let mut events: Vec<(i32, i32, i32)> = Vec::new();
    for &(l, r, h) in buildings {
        events.push((l, h, r));    // start
        events.push((r, 0, 0));    // end
    }
    events.sort_by(|a, b| {
        if a.0 != b.0 { a.0.cmp(&b.0) } else { b.1.cmp(&a.1) }  // taller starts first
    });
    let mut heap: BinaryHeap<(i32, i32)> = BinaryHeap::new();
    heap.push((0, i32::MAX));
    let mut result = Vec::new();
    for (x, h, r) in events {
        if h != 0 { heap.push((h, r)); }
        while let Some(&(_, end)) = heap.peek() {
            if end <= x { heap.pop(); } else { break; }
        }
        let cur_h = heap.peek().unwrap().0;
        match result.last() {
            Some(&(_, prev)) if prev == cur_h => {}
            _ => result.push((x, cur_h)),
        }
    }
    result
}

pub fn max_overlap(intervals: &[(i32, i32)]) -> i32 {
    let mut events: Vec<(i32, i32)> = Vec::new();
    for &(s, e) in intervals {
        events.push((s, 1));
        events.push((e, -1));
    }
    events.sort_by(|a, b| {
        if a.0 != b.0 { a.0.cmp(&b.0) } else { b.1.cmp(&a.1) }  // +1 before -1
    });
    let (mut cur, mut best) = (0, 0);
    for (_, d) in events {
        cur += d;
        if cur > best { best = cur; }
    }
    best
}

fn main() {
    let sk = skyline(&[(2,9,10),(3,7,15),(5,12,12),(15,20,10),(19,24,8)]);
    print!("Skyline:");
    for (x, h) in &sk { print!(" ({},{})", x, h); }
    println!();
    println!("Max overlap: {}", max_overlap(&[(1,5),(2,6),(3,4),(7,8)]));
}

// NOTES
// -----
// Differences from Java:
//   * Java's geometry.java does not include sweep-line algorithms; we add
//     skyline and overlap classics here.
//   * BinaryHeap is max-heap by default — convenient for the skyline.
//   * Tuple ordering / sort_by with closures keeps the comparator concise.
